src/Generator/Passes/DelegatesPass.cs

```diff
@@ -119,61 +213,111 @@ private void AddDelegatesToDictionary(Declaration decl, Module module)
+                }
+                
+                var typeDefType1 = new TypedefType
+                {
+                    Declaration = @delegate
+                };
+                
+                if(!Context.Delegates.ContainsKey(decl))
+                    Context.Delegates.Add(decl, existingDelegate);
+                return new QualifiedType(typeDefType1);
             }
 
-            existingDelegate = new DelegateDefinition(module.OutputNamespace, delegateString);
+            existingDelegate = new DelegateDefinition(module.OutputNamespace, delegateString, access, callingConvention);
             Context.Delegates.Add(decl, existingDelegate);
 
-            libsDelegates[module].Add(delegateString, existingDelegate);
+            if(libsDelegates.ContainsKey(module))
+                libsDelegates[module].Add(delegateString, existingDelegate);
 
             namespaceDelegates.Declarations.Add(@delegate);
+            return new QualifiedType(typeDefType);
         }
 
-        private void VisitFunctionTypeParameters(Function function)
+        private static TypedefDecl FindMatchingTypedef(IEnumerable<TypedefDecl> typedefs, FunctionType functionType)
         {
-            foreach (var param in from param in function.Parameters
-                                  where !(param.Type is TypedefType)
-                                  let paramType = param.Type.Desugar()
-                                  where paramType.IsAddress() &&
-                                      paramType.GetPointee() is FunctionType
-                                  select param)
-            {
-                var module = function.TranslationUnit.Module;
-                AddDelegatesToDictionary(param, module);
-            }
+            return (from typedef in typedefs
+                    let type = (FunctionType)typedef.Type.GetPointee()
+                    where type.ReturnType == functionType.ReturnType &&
+                          type.Parameters.SequenceEqual(functionType.Parameters, ParameterTypeComparer.Instance)
+                    select typedef).SingleOrDefault();
         }
 
         public override bool VisitFunctionDecl(Function function)
         {
             if (!base.VisitFunctionDecl(function) || function.Ignore)
                 return false;
 
-            // HACK : We should shift this call to VisitFunctionTypeParameters in VisitParameter. We can't do it now
-            // because we need to use the module and a since a parameter's namespace is null, we can't reach the
-            // translation unit and thus the module
-            VisitFunctionTypeParameters(function);
-
+            function.ReturnType = CheckType(function.Namespace, function.ReturnType, function);
+            foreach (var parameter in function.Parameters)
+            {
+                parameter.QualifiedType = CheckType(function.Namespace, parameter.QualifiedType, parameter);
```

Comment:
You set this once and then set it another way. Use the same method if possible. If not, use an else.